### PR TITLE
r/ssm_maint_windows_task - add CloudWatch config + plan time validations

### LINKED ISSUE #17554: AWS SSM Maintenance Window Task Outputs allow cloudwatch in the console, no option in terraform
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

aws_ssm_maintenance_window_task does not have any parameters for cloudwatch output for Run Commands. This is an option in the console but I can't seem to use it with my terraform.

### New or Affected Resource(s)

* aws_ssm_maintenance_window_task

### Potential Terraform Configuration
```hcl 
resource "aws_ssm_maintenance_window_task" "test" {
  window_id        = aws_ssm_maintenance_window.test.id
  task_type        = "RUN_COMMAND"
  task_arn         = "AWS-RunShellScript"
  priority         = 1
  service_role_arn = aws_iam_role.test.arn
  max_concurrency  = "2"
  max_errors       = "1"

  targets {
    key    = "WindowTargetIds"
    values = [aws_ssm_maintenance_window_target.test.id]
  }

  task_invocation_parameters {
    run_command_parameters {
      cloud_watch_output_config {
        cloud_watch_log_group_name = "example"
        cloud_watch_output_enabled   = true
      }
    }
  }
}
```

### References

* https://docs.aws.amazon.com/systems-manager/latest/APIReference/API_MaintenanceWindowRunCommandParameters.html#systemsmanager-Type-MaintenanceWindowRunCommandParameters-CloudWatchOutputConfig
* https://docs.aws.amazon.com/sdk-for-go/api/service/ssm/#MaintenanceWindowRunCommandParameters

### PR DESCRIPTION
**Pull request by DrFaust92**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates #13624
Relates #13826
Closes #17554

Release note for [CHANGELOG](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md):

```release-note
resource_aws_ssm_maintance_window_task: add `cloudwatch_config` argument.
resource_aws_ssm_maintance_window_task: add plan time validation to `task_type`, `service_role_arn`, `targets.notification_arn`, `targets.service_role_arn`.
```

Output from acceptance testing:

```
$ make testacc TESTARGS='-run=TestAccAWSSSMMaintenanceWindowTask_'
--- PASS: TestAccAWSSSMMaintenanceWindowTask_basic (115.08s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource (100.42s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters (139.40s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters (95.09s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters (140.03s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParametersCloudWatch (149.73s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters (68.90s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_emptyNotificationConfig (59.16s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_disappears (57.53s)
```

### CODE REVIEW

**Comment by giuliocalzolari** [comment]
Do you have ETA to merge it?

**Comment by giuliocalzolari** [comment]
@slapula do you know when will be merged?

**Comment by DrFaust92** [comment]
having trouble with CW test, cant find where i messed up at the moment.

**Comment by DrFaust92** [comment]
Rebased

```
--- PASS: TestAccAWSSSMMaintenanceWindowTask_basic (115.08s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource (100.42s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters (139.40s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters (95.09s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters (140.03s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParametersCloudWatch (149.73s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters (68.90s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_emptyNotificationConfig (59.16s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_disappears (57.53s)
```

File: .changelog/11774.txt [hunk 1]
```diff
@@ -0,0 +1,12 @@
+```release-note:enhancement
+resource/aws_ssm_maintenance_window_task: Add `cloudwatch_config` argument
+```
+
+```release-note:enhancement
+resource/aws_ssm_maintenance_window_task: Add `document_version` argument
```
Comment (bflad) [inline .changelog/11774.txt:6]:
``````suggestion
resource/aws_ssm_maintenance_window_task: Add `task_invocation_parameters` `run_command_parameters` block `cloudwatch_config` and `document_version` arguments
``````

File: .changelog/11774.txt [hunk 2]
```diff
@@ -0,0 +1,12 @@
...
+```
+
+```release-note:enhancement
+resource/aws_ssm_maintenance_window_task: Add plan time validation to `task_type`, `service_role_arn`, `targets.notification_arn`, `targets.service_role_arn`, `max_concurrency`, `max_errors`, `targets`, `priority`, 
+`task_invocation_parameters.run_command_parameters.comment`, `task_invocation_parameters.run_command_parameters.document_hash`, `task_invocation_parameters.run_command_parameters.timeout_seconds`, `task_invocation_parameters.run_command_parameters.notification_config.notification_events`
```
Comment (bflad) [inline .changelog/11774.txt:11]: Newlines should not be added in CHANGELOG entries. Reference: https://github.com/hashicorp/terraform-provider-aws/issues/17135

```suggestion
resource/aws_ssm_maintenance_window_task: Add plan time validation to `max_concurrency`, `max_errors`, `priority`, `service_role_arn`, `targets`, `targets.notification_arn`, `targets.service_role_arn`, `task_type`, `task_invocation_parameters.run_command_parameters.comment`, `task_invocation_parameters.run_command_parameters.document_hash`, `task_invocation_parameters.run_command_parameters.timeout_seconds`, and `task_invocation_parameters.run_command_parameters.notification_config.notification_events` arguments
```

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 3]
```diff
@@ -31,19 +31,22 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
 			},
 
 			"max_concurrency": {
-				Type:     schema.TypeString,
-				Required: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile(`^([1-9][0-9]*|[1-9][0-9]%|[1-9]%|100%)$`), ""),
```
Comment (bflad) [inline aws/resource_aws_ssm_maintenance_window_task.go:36]: `validation.StringMatch()` errors should include a human friendly message over the raw regular expression. Reference: https://github.com/hashicorp/terraform-provider-aws/issues/16628

```suggestion
				ValidateFunc: validation.StringMatch(regexp.MustCompile(`^([1-9][0-9]*|[1-9][0-9]%|[1-9]%|100%)$`), "must be a number without leading zeros or a percentage between 1% and 100% without leading zeros and ending with the percentage symbol"),
```

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 4]
```diff
@@ -31,19 +31,22 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
 			},
 
 			"max_errors": {
-				Type:     schema.TypeString,
-				Required: true,
+				Type:         schema.TypeString,
+				Required:     true,
+				ValidateFunc: validation.StringMatch(regexp.MustCompile(`^([1-9][0-9]*|[0]|[1-9][0-9]%|[0-9]%|100%)$`), ""),
```
Comment (bflad) [inline aws/resource_aws_ssm_maintenance_window_task.go:42]:
```suggestion
				ValidateFunc: validation.StringMatch(regexp.MustCompile(`^([1-9][0-9]*|[0]|[1-9][0-9]%|[0-9]%|100%)$`), "must be zero, a number without leading zeros, or a percentage between 1% and 100% without leading zeros and ending with the percentage symbol"),
```

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 5]
```diff
@@ -165,22 +173,26 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
-											ssm.DocumentHashTypeSha1,
-										}, false),
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validation.StringInSlice(ssm.DocumentHashType_Values(), false),
+									},
+									"document_version": {
+										Type:         schema.TypeString,
+										Optional:     true,
+										ValidateFunc: validation.StringMatch(regexp.MustCompile(`([$]LATEST|[$]DEFAULT|^[1-9][0-9]*$)`), ""),
```
Comment (bflad) [inline aws/resource_aws_ssm_maintenance_window_task.go:195]:
```suggestion
										ValidateFunc: validation.StringMatch(regexp.MustCompile(`([$]LATEST|[$]DEFAULT|^[1-9][0-9]*$)`), "must be $DEFAULT, $LATEST, or a version number"),
```

File: aws/resource_aws_ssm_maintenance_window_task.go [hunk 6]
```diff
@@ -190,23 +202,31 @@ func resourceAwsSsmMaintenanceWindowTask() *schema.Resource {
...
+													Elem: &schema.Schema{
+														Type: schema.TypeString,
+														ValidateFunc: validation.StringInSlice([]string{
+															"All",
+															"InProgress",
+															"Success",
+															"TimedOut",
+															"Cancelled",
+															"Failed",
+														}, false),
```
Comment (bflad) [inline aws/resource_aws_ssm_maintenance_window_task.go:222]: Looks like the enumeration is now available:

```suggestion
														Type:         schema.TypeString,
														ValidateFunc: validation.StringInSlice(ssm.NotificationEvent_Values(), false),
```

**Review by bflad — COMMENTED** [review]
Just some touch up before verifying ✨

**Review by bflad — APPROVED** [review]
Took me a bit to get through all the validation updates, but looks good to me. 🚀 

Output from acceptance testing in AWS Commercial:

```
--- PASS: TestAccAWSSSMMaintenanceWindowTask_disappears (19.58s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_emptyNotificationConfig (19.85s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters (22.39s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource (34.36s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_basic (35.48s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters (42.68s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParametersCloudWatch (44.70s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters (45.14s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters (46.92s)
```

Output from acceptance testing in AWS GovCloud (US):

```
--- PASS: TestAccAWSSSMMaintenanceWindowTask_disappears (24.63s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_emptyNotificationConfig (24.80s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationStepFunctionParameters (27.60s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationLambdaParameters (39.96s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_basic (41.36s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_updateForcesNewResource (41.58s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationAutomationParameters (48.45s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParameters (49.11s)
--- PASS: TestAccAWSSSMMaintenanceWindowTask_TaskInvocationRunCommandParametersCloudWatch (54.92s)
```

**Pull request merged**